packages/tools/src/eventListeners/segmentation/stackImageChangeEventListener.ts

```diff
@@ -0,0 +1,213 @@
+ *
+ * - First we get all tools which are active, passive or enabled on the element.
+ * - If any of these tools have a `renderAnnotation` method, then we render them.
+ * - Note that these tools don't necessarily have to be instances of  `AnnotationTool`,
+ *   Any tool may register a `renderAnnotation` method (e.g. a tool that displays an overlay).
+ *
+ * @param evt - The normalized IMAGE_RENDERED event.
+ */
+function _stackImageChangeEventListener(evt) {
+  const eventData = evt.detail;
+  const { viewportId, renderingEngineId } = eventData;
+  const { viewport } = getEnabledElementByIds(
+    viewportId,
+    renderingEngineId
+  ) as { viewport: Types.IStackViewport };
+
+  const toolGroup = getToolGroupForViewport(viewportId, renderingEngineId);
+  let toolGroupSegmentationRepresentations =
+    SegmentationState.getSegmentationRepresentations(toolGroup.id) || [];
+
+  toolGroupSegmentationRepresentations =
+    toolGroupSegmentationRepresentations.filter(
+      (representation) => representation.type === Representations.Labelmap
+    );
+
+  if (!toolGroupSegmentationRepresentations?.length) {
+    return;
+  }
+
+  const segmentationRepresentations = {};
+  toolGroupSegmentationRepresentations.forEach((representation) => {
+    const segmentation = SegmentationState.getSegmentation(
+      representation.segmentationId
+    );
+
+    if (!segmentation) {
+      return;
+    }
+
+    const labelmapData =
+      segmentation.representationData[Representations.Labelmap];
+
+    if (isVolumeSegmentation(labelmapData)) {
+      return;
+    }
+
+    const { referencedImageIds, imageIds } =
+      labelmapData as LabelmapSegmentationDataStack;
+
+    segmentationRepresentations[representation.segmentationRepresentationUID] =
+      {
+        referencedImageIds,
+        segmentationImageIds: imageIds,
+      };
+  });
+
+  const representationList = Object.keys(segmentationRepresentations);
+  const currentImageId = viewport.getCurrentImageId();
+  const actors = viewport.getActors();
+
+  actors.forEach((actor) => {
+    if (representationList.includes(actor.uid)) {
+      const segmentationActor = actor.actor;
+
+      const { referencedImageIds, segmentationImageIds } =
+        segmentationRepresentations[actor.uid];
+
+      const derivedImageId = getDerivedImageId(
+        currentImageId,
+        referencedImageIds,
+        segmentationImageIds
+      );
+
+      const segmentationImageData = segmentationActor
+        .getMapper()
+        .getInputData();
+
+      const derivedImage = cache.getImage(derivedImageId);
+
+      const { origin, dimensions, spacing, direction } =
+        viewport.getImageDataMetadata(derivedImage);
+
+      segmentationImageData.setOrigin(origin);
+      segmentationImageData.modified();
+
+      if (segmentationImageData.getDimensions()[0] !== dimensions[0]) {
+        // IMPORTANT: Not sure why we can't just update the dimensions
```

Comment:
The defect you were seeing is because the other version of it uses the primary image ID for the data, and this one, correctly uses the derivedImageId instead.  If you fix the bug in the other spot, you should be able to update the dimensions and call modified.

Reply:
No it is not about that, it is about we have a stack of CT and PT and when we change the slice (from CT to PT) we can't use the CT actor to fill the PT segmentation, we should re-create based on its new metadata